### ISSUE TITLE
Rework the FX Module

### ISSUE THREAD

**Issue by saivan** [svgdotjs/svg.js#776]
I was wondering whether it would make sense to batch the animation frames for the fx module, like I did with the svg.declarative.js plugin. 

I wrote a [nice bit of code here](https://github.com/saivan/svg.declarative.js/blob/master/source/drawloop.js) that has the same api as the window's `requestAnimationFrame` and `setTimeout` functions. Of course I only implemented what I needed, but I saw a pretty huge speedup because there wasn't any setup or teardown required by the function calls.

Let me know what you all think?
Are there any good reasons not to do this?

**Comment by Fuzzyma** [comment]
Is your request animation frame always running in the background? Interesting idea. So you just attach an animation to your running requestAnimationFrame whenever needed.
Jep that sounds like it could speedup stuff a lot!

**Comment by saivan** [comment]
Yep thats what I do! It doesn't actually always run, I keep a queue (implemented as a linked list so all of the operations are O(1)) and if the queue is not empty, I request another animation frame.

**Comment by saivan** [comment]
# NOTES ON FX

- We should remove functions that are related to a queue, and just use a queue implementation instead. Propose SVG.Queue Which I have already implemented for declarative: This should remove the functions:
  - first
  - last
  - add

- There is no need to keep the queue ordered in time. Just have a simple method that checks if members of the queue are executing; this will be super fast, but it will allow for the proposals below.

- If we implement a queue as a linked list called `pending`, adding items from the front to the back is a very fast operation. So we can just check every pending situation to see if it is supposed to be running, if it is, run step on it and if it isn't, add it to the back of the queue.


# QUESTIONS ON FX

- What is target supposed to do? Could we improve that?
- Could we rename Situation to Animation? Why did we use Situation?
- We should make animation in charge of actually performing the animations and let fx focus on timelining
- Maybe we should rename FX to Timeline


# PROPOSED FX ARCHITECTURE
As a main change, we would like to be able to make animations run concurrently, right now, this is impossible since the current timeline model is:

```
  || Declared at this time
  || -----t1-----|===t2===|----t3-----|====t4====|-----t5------|==========>
  ||      A1         D1        A2          D2           A3
```

And the syntax to create this right now is:

```javascript
  el.animate(t1, "-", 0) // Declare A1
      TARGET A1
    .delay(t2)
    .animate(t3)
      TARGET A2
    .animate(t5, "-", t3)
      TARGET A3
```

Whereas it would be better to have the ability to have concurrent timelines running so that people can start to architect more useful animations:

```
  || Both declared Here
  ||--------t1----------|==t2==|-----t3-----|
  ||        A1             D1         A2
  ||==t4==|----------t5----------|====t6====|-----t7-------|
      D3             A3               D4           A4
```

And the API would look like

```javascript
  el.animate(t1, '-', 0)
      TARGET A1
    .animate(t3, '-', t2)
      TARGET A2
    .animate(t5, '-', t4, true) // This parameter restarts the timer
      TARGET A3
    .delay(t6)
    .animate(t7, '-')
      TARGET A4
```

The only **breaking change** here would be, that you could achieve this exact timeline layout with this code:

```javascript
  el.animate(t1, '-', 0)
      TARGET A1
    .animate(t3, '-', t2)
      TARGET A2

  el.animate(t5, '-', t4, true)
      TARGET A3
    .animate(t7, '-', t6)
      TARGET A4
```

Whereas the old code would have just run A3 and A4 after A1 and A2. This is super useful as it would allow for crazier animations, for example right now its impossible to do something as simple as making a square fade in and while its fading in, make it grow in size.

```
  || NOW
  ||-------------------t1---------------------|
  ||         Make Square fade in
  ||                 |-------t2--------|
  ||                     Make it BIG!
```

# PROPOSED API

FX:

  User Methods
  ------------

    Creating situations
    -------------------
    animate (duration, ease, delay, epoch)
      Creates a new situation and adds it to the queue. If epoch is true, the
      epoch time for new animations gets set to the current time.

    delay (by)
      Advances the epoch time for new animations by the given time.

    loop (times, reverse)
      Sets the number of times to loop the current animation and whether it
      should be allowed to reverse itself

    Modifying playback
    ------------------
    play ()
      Plays if paused

    pause ()
      Pauses the animation

    speed (newSpeed)
      changes the speed of the animation

    seek (dt)
      Advances (or moves back) the time by the time given

    persist (dt || forever)
        Tells us how long to keep an animation in the history before deleting it

    finish ()
      Jumps to the end of all of the animations

    Injecting User Code
    -------------------
    Users can provide an optional name to events if they would like to cancel
    them. To cancel a particular event, just cancel the event by its name. If
    the user provides a duplicate name, they will only be able to cancel the
    last event by that name (possibly a useful debouncing feature).

    cancel (name)
      Cancels an event by name

    start (event, name)
      Bind an event to run once to a situation

    during (event, name)
      Bind an event to run during every frame that an animation executes

    end (event, name)
      Bind an event to run after an animation ends

    while (event, name)
      Bind an event to run if there are situations left, we have to allow for
      more than one of these.

    conclude (event, name)
      Bind an event to run once there are no situations left to run

    Getting Information
    -------------------
    progress ()
      Tells the user how much of the animation is complete it:
        - returns an object that has references to the running animations
        - returns the last time that an animation ran

  Hidden Methods
  --------------
    _step(dt)
      Steps the animation by the time dt

    _continue()
      Continues the animation if it has been laying dormant


# Questions

- Should we allow users to scrub back to undo animations that have already completed? If so, we can keep a history queue and a running queue to keep things as fast as possible (but the memory footprint increases)

**Comment by Fuzzyma** [comment]
I want to add that we agreed on removing the `after, during...` functions in favour for events.

Also I would like to introduce a better concept for relative Numbers.
When we introduced them for `animation().dmove()` it was in a very hackish way.
So I would like to refine these. The `SVG.Number` constructor would look like this:

```js
new SVG.Number(x, relative)
```

The different behavior to normal numbers is already implemented. Only the morph function changed such that it sets a destination which is relative from the value of the number.

**Comment by saivan** [comment]
After some discussions yesterday, we decided that ease should no longer be a parameter of animate, it is more like a property of the timeline in question, so the new api for animate would be:

```
element.animate(duration, delay, now)
```

There would also be shortcuts:

```
element.duration(newDuration)
// or
element.delay(newDelay, now)
```

**Comment by saivan** [comment]
This has been merged into 3.0
